src/run.cpp

```diff
@@ -0,0 +1,192 @@
+    g_queue = nullptr;
+
+    if (g_pmap)
+        delete g_pmap;
+    g_pmap = nullptr;
+
+    int v = 0;
+    if (g_verbosity.count("loud"))
+        v = 1000;
+    else if (g_verbosity.count("progress"))
+        v = 1;
+
+    if(!logger) {
+        if(log_fname)
+            logger = new Logger(c, nrules, v, log_fname, freq);
+        else {
+            logger = new PyLogger();
+            logger->setVerbosity(v);
+        }
+    }
+
+    g_init = timestamp();
+    char run_type[BUFSZ];
+    strcpy(run_type, "LEARNING RULE LIST via ");
+    char const *type = "node";
+    if (curiosity_policy == 1) {
+        strcat(run_type, "CURIOUS");
+        g_queue = new Queue(curious_cmp, run_type);
+        type = "curious";
+    } else if (curiosity_policy == 2) {
+        strcat(run_type, "LOWER BOUND");
+        g_queue = new Queue(lb_cmp, run_type);
+    } else if (curiosity_policy == 3) {
+        strcat(run_type, "OBJECTIVE");
+        g_queue = new Queue(objective_cmp, run_type);
+    } else if (curiosity_policy == 4) {
+        strcat(run_type, "DFS");
+        g_queue = new Queue(dfs_cmp, run_type);
+    } else {
+        strcat(run_type, "BFS");
+        g_queue = new Queue(base_cmp, run_type);
+    }
+
+    if (map_type == 1) {
+        strcat(run_type, " Prefix Map\n");
+        PrefixPermutationMap* prefix_pmap = new PrefixPermutationMap;
+        g_pmap = (PermutationMap*) prefix_pmap;
+    } else if (map_type == 2) {
+        strcat(run_type, " Captured Symmetry Map\n");
+        CapturedPermutationMap* cap_pmap = new CapturedPermutationMap;
+        g_pmap = (PermutationMap*) cap_pmap;
+    } else {
+        strcat(run_type, " No Permutation Map\n");
+        NullPermutationMap* null_pmap = new NullPermutationMap;
+        g_pmap = (PermutationMap*) null_pmap;
+    }
+
+    g_tree = new CacheTree(nsamples, nrules, c, rules, labels, meta, ablation, calculate_size, type);
+    if (g_verbosity.count("progress"))
+        printf("%s", run_type);
+
+    bbound_begin(g_tree, g_queue);
+
+    return 0;
+}
+
+int run_corels_loop(size_t max_num_nodes) {
+    if((g_tree->num_nodes() < max_num_nodes) && !g_queue->empty()) {
+        bbound_loop(g_tree, g_queue, g_pmap);
+        return 0;
+    }
+    return -1;
+}
+
+double run_corels_end(int** rulelist, int* rulelist_size, int** classes, int early, int latex_out, rule_t* rules, rule_t* labels, char* opt_fname)
+{
+    bbound_end(g_tree, g_queue, g_pmap, early);
+
+    const tracking_vector<unsigned short, DataStruct::Tree>& r_list = g_tree->opt_rulelist();
+    const tracking_vector<bool, DataStruct::Tree>& preds = g_tree->opt_predictions();
+
+    double accuracy = 1.0 - g_tree->min_objective() + g_tree->c() * r_list.size();
+
+    *rulelist = (int*)malloc(sizeof(int) * r_list.size());
+    *classes = (int*)malloc(sizeof(int) * (1 + r_list.size()));
+    *rulelist_size = r_list.size();
+    for(size_t i = 0; i < r_list.size(); i++) {
+        (*rulelist)[i] = r_list[i];
```

Comment:
rulelist->i should be equivalent

Reply:
Hmm, that doesn't seem to work for me - it gives an error:
error: request for member ‘i’ in ‘* rulelist’, which is of non-class type ‘int*’

Reply:
This will change if we make this a vector (which I think we should)